File: sdk/python/agentguard/_base_url.py
"""Validation for credential-bearing AgentGuard deployment URLs."""

from __future__ import annotations

from urllib.parse import urlsplit


def normalize_base_url(base_url: str) -> str:
    """Return a normalized absolute HTTP(S) deployment URL.

    AgentGuard clients attach an organization credential to requests. Requiring
    and validating the destination prevents an omitted or relative URL from
    silently sending that credential to an unintended host.
    """
    value = base_url.strip().rstrip("/")
    parsed = urlsplit(value)
    if (
        parsed.scheme not in {"http", "https"}
        or not parsed.netloc
        or parsed.username is not None
        or parsed.password is not None
        or parsed.path not in {"", "/"}
        or parsed.query
        or parsed.fragment
    ):
        raise ValueError(
            "base_url must be an absolute http(s) AgentGuard deployment origin"
        )
    if parsed.scheme == "http" and parsed.hostname not in {
        "localhost",
        "127.0.0.1",
        "::1",
    }:
        raise ValueError(
            "base_url must use https unless the AgentGuard deployment is on loopback"
        )
    return value

File: sdk/node/src/base-url.ts
/** Validate and normalize an explicit AgentGuard deployment origin. */
export function normalizeBaseUrl(baseUrl: string): string {
  let parsed: URL;
  try {
    parsed = new URL(baseUrl.trim().replace(/\/+$/, ''));
  } catch {
    throw new TypeError(
      'baseUrl must be an absolute HTTP(S) AgentGuard deployment origin',
    );
  }

  if (
    !['http:', 'https:'].includes(parsed.protocol) ||
    parsed.username ||
    parsed.password ||
    (parsed.pathname !== '/' && parsed.pathname !== '') ||
    parsed.search ||
    parsed.hash
  ) {
    throw new TypeError(
      'baseUrl must be an absolute HTTP(S) AgentGuard deployment origin',
    );
  }

  const loopbackHosts = new Set(['localhost', '127.0.0.1', '[::1]']);
  if (parsed.protocol === 'http:' && !loopbackHosts.has(parsed.hostname)) {
    throw new TypeError(
      'baseUrl must use HTTPS unless the AgentGuard deployment is on loopback',
    );
  }

  return parsed.origin;
}
